src/main/java/org/mskcc/cmo/messaging/impl/JSGatewayImpl.java

```diff
@@ -0,0 +1,243 @@
+import org.mskcc.cmo.common.FileUtil;
+import org.mskcc.cmo.messaging.Gateway;
+import org.mskcc.cmo.messaging.MessageConsumer;
+import org.mskcc.cmo.messaging.utils.SSLUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Primary;
+import org.springframework.stereotype.Component;
+
+@Primary
+@Component
+public class JSGatewayImpl implements Gateway {
+    @Value("${nats.tls_channel:false}")
+    private boolean tlsChannel;
+
+    @Value("${nats.url}")
+    private String natsURL;
+
+    @Value("${metadb.publishing_failures_filepath}")
+    private String metadbPubFailuresFilepath;
+
+    private FileUtil fileUtil;
+
+    private File pubFailuresFile;
+
+    @Autowired
+    private void initPubFailuresFile() throws IOException {
+        this.pubFailuresFile = fileUtil.getOrCreateFileWithHeader(
+                metadbPubFailuresFilepath, PUB_FAILURES_FILE_HEADER);
+    }
+
+    @Autowired
+    SSLUtils sslUtils;
+
+    private static final String PUB_FAILURES_FILE_HEADER = "DATE\tTOPIC\tMESSAGE\n";
+    private Connection natsConnection;
+    private JetStream jsConnection;
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final Map<String, JetStreamSubscription> subscribers = new HashMap<String,
+            JetStreamSubscription>();
+    private volatile boolean shutdownInitiated;
+    private final ExecutorService exec = Executors.newSingleThreadExecutor();
+    private final CountDownLatch publishingShutdownLatch = new CountDownLatch(1);
+    private final BlockingQueue<PublishingQueueTask> publishingQueue =
+        new LinkedBlockingQueue<PublishingQueueTask>();
+    private final Log LOG = LogFactory.getLog(StanGatewayImpl.class);
+
+    private class PublishingQueueTask {
+        String topic;
+        Object message;
+
+        PublishingQueueTask(String topic, Object message) {
+            this.topic = topic;
+            this.message = message;
+        }
+    }
+
+    private class NATSPublisher implements Runnable {
+
+        Connection natsConn;
+        JetStream jetStreamConn;
+        boolean interrupted = false;
+
+        NATSPublisher() throws Exception {
+            JetStreamOptions jsOptions = new JetStreamOptions.Builder()
+                    .build();
+            natsConn = Nats.connect(natsConnection.getOptions());
+            this.jetStreamConn = natsConn.jetStream(jsOptions);
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    PublishingQueueTask task = publishingQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (task != null) {
+                        String msg = mapper.writeValueAsString(task.message);
+                        try {
+                            jetStreamConn.publishAsync(task.topic, msg.getBytes(StandardCharsets.UTF_8));
```

Comment:
does this need to use the `publishAsync()` or can it use the `publish()` method like the other streaming connection impl

Reply:
notes: returns completable future of the publish acknowledgement? 

@divyamadala30 can you add some details. here?

Reply:
Ben: with workers operating with their own threads then we can prob publish synchronously